cranelift/codegen/src/isa/x64/inst/regs.rs

```diff
@@ -0,0 +1,261 @@
+//! Registers, the Universe thereof, and printing.
+//!
+//! These are ordered by sequence number, as required in the Universe.  The strange ordering is
+//! intended to make callee-save registers available before caller-saved ones.  This is a net win
+//! provided that each function makes at least one onward call.  It'll be a net loss for leaf
+//! functions, and we should change the ordering in that case, so as to make caller-save regs
+//! available first.
+//!
+//! TODO Maybe have two different universes, one for leaf functions and one for non-leaf functions?
+//! Also, they will have to be ABI dependent.  Need to find a way to avoid constructing a universe
+//! for each function we compile.
+
+use alloc::vec::Vec;
+use std::string::String;
+
+use regalloc::{RealReg, RealRegUniverse, Reg, RegClass, RegClassInfo, NUM_REG_CLASSES};
+
+use crate::machinst::pretty_print::ShowWithRRU;
+use crate::settings;
+
+// Hardware encodings for a few registers.
+
+pub const ENC_RBX: u8 = 3;
+pub const ENC_RSP: u8 = 4;
+pub const ENC_RBP: u8 = 5;
+pub const ENC_R12: u8 = 12;
+pub const ENC_R13: u8 = 13;
+pub const ENC_R14: u8 = 14;
+pub const ENC_R15: u8 = 15;
+
+fn gpr(enc: u8, index: u8) -> Reg {
+    Reg::new_real(RegClass::I64, enc, index)
+}
+
+fn r12() -> Reg {
+    gpr(ENC_R12, 0)
+}
+fn r13() -> Reg {
+    gpr(ENC_R13, 1)
+}
+pub fn r14() -> Reg {
+    gpr(ENC_R14, 2)
+}
+fn r15() -> Reg {
+    gpr(ENC_R15, 3)
+}
+fn rbx() -> Reg {
+    gpr(ENC_RBX, 4)
+}
+pub fn rsi() -> Reg {
+    gpr(6, 5)
+}
+pub fn rdi() -> Reg {
+    gpr(7, 6)
+}
+pub fn rax() -> Reg {
+    gpr(0, 7)
+}
+pub fn rcx() -> Reg {
+    gpr(1, 8)
+}
+pub fn rdx() -> Reg {
+    gpr(2, 9)
+}
+pub fn r8() -> Reg {
+    gpr(8, 10)
+}
+pub fn r9() -> Reg {
+    gpr(9, 11)
+}
+fn r10() -> Reg {
+    gpr(10, 12)
+}
+fn r11() -> Reg {
+    gpr(11, 13)
+}
+
+fn fpr(enc: u8, index: u8) -> Reg {
+    Reg::new_real(RegClass::V128, enc, index)
+}
+fn xmm0() -> Reg {
+    fpr(0, 14)
+}
```

Comment:
```suggestion
fn xmm(num: u8) -> Reg {
    fpr(num, 14 + num)
}
```

Reply:
> it may require a bunch of changes all over the place.

These functions are not exported from this module.

> I'll punt on this one, it does break the symmetry with GPR

Fair enough